### PR TITLE
Add /rest/tree API call

### PR DESCRIPTION
**Pull request by AudriusButkevicius**
We'll need this for the selective sync, plus useful to have in general, as other can query this.

### CODE REVIEW

File: internal/model/model.go [hunk 1]
```diff
@@ -1388,6 +1388,83 @@ func (m *Model) RemoteLocalVersion(folder string) int64 {
 	return ver
 }
 
+func (m *Model) GetDirectoryTree(folder, prefix string, levels int, dirsonly bool) map[string]interface{} {
```
Comment (calmh) [inline internal/model/model.go]: Looks sane enough, but could we have a test just for the sake of it? I isn't completely obvious to me that the stack popping mechanism etc works in all cases like `foo/bar/baz` followed directly by `something/else/entirely` and so on...

Comment (AudriusButkevicius) [inline internal/model/model.go]: Makes sense.

**Comment by AudriusButkevicius** [comment]
So this relies on the fact that if we are at `some/file/here`  we must have seen both: `some/file/` and `some/`.

Not sure if that is a correct assumption?

**Comment by calmh** [comment]
Should be safe I think... Perhaps it would be worth benchmarking to see how fast/slow it is, but otherwise just splitting the string on `/` and walking the map for each entry would be safe and simple.

**Comment by AudriusButkevicius** [comment]
I haven't forgotten this, just been busy/lazy.

**Comment by AudriusButkevicius** [comment]
```
BenchmarkTreeStack10000_50            10         154319740 ns/op
BenchmarkTreeMap__10000_50             5         271634520 ns/op
BenchmarkTreeStack10000_10            20          70307705 ns/op
BenchmarkTreeMap__10000_10            20          96613720 ns/op
BenchmarkTreeStack00100_50           100          12231559 ns/op
BenchmarkTreeMap__00100_50           100          13411698 ns/op
BenchmarkTreeStack00100_10           100          11921275 ns/op
BenchmarkTreeMap__00100_10           100          12411712 ns/op
```
Where 10000 is the number of index entries, 50 is how deep the file is.
Each directory before the file counts as an index entry.

**Comment by calmh** [comment]
```
BenchmarkTree_10000_50	       5	 241859226 ns/op	127420668 B/op	  618834 allocs/op
```

That's a fairly expensive operation :)

**Comment by AudriusButkevicius** [comment]
Ideas on improving it?
I still have the stack approach, but that one goes nuts when you have missing tree leaves in the index.

**Comment by AudriusButkevicius** [comment]
I wonder how this works with 1.9M files that someone has.

**Comment by dzarda** [comment]
Personally, I'm a bit afraid about the memory requirements of this thing... But perhaps this is just my _performance-obsessed_ alter ego :)

**Comment by AudriusButkevicius** [comment]
Well if you want to fit the whole tree, I don't see how you can do it differently though.
Hence why the levels/prefix thing is there.

**Comment by calmh** [comment]
No, if we want to return it as a map this is what we get - maps are (fairly) expensive and that's a lot of them. So that's fine. It's just something to be aware of. (A pre-allocated `[]SomeStructType` would be more *efficient*, but more annoying to use from the client side as well. But it's unclear what the client side is yet, on the other hand. :)

**Pull request merged**